File: Python/ByTier/E/2147_E_Maximum_OR_Popcount.py
# Problem: CF 2147 E - Maximum OR Popcount
# https://codeforces.com/contest/2147/problem/E

"""
E. Maximum OR Popcount
Purpose:
    Given an array of non-negative integers and multiple queries, each query allows
    performing at most 'b_i' operations (incrementing elements by 1). The goal is
    to maximize the number of bits equal to 1 in the bitwise OR of all numbers.
Algorithms:
    - For each bit position from high to low, calculate how much cost is needed to set that bit
      in the OR result. Greedily decide how many operations can be spent to gain set bits.
    - Use prefix/suffix techniques combined with bit manipulation and binary search principles.
Time Complexity: O(n * 32 + q * 32) where n is array size and q is number of queries
Space Complexity: O(1) excluding input storage and output arrays
"""

import sys
from bisect import bisect_right
from random import getrandbits

rand1 = getrandbits(32)
rand2 = getrandbits(32)
if rand1 == rand2:
    rand2 = 2**32 // rand2


class Wrap(int):
    def __init__(self, x):
        int.__init__(x)

    def __hash__(self):
        return super(Wrap, self).__hash__() ^ rand1 ^ rand2


def inp():
    return sys.stdin.readline().strip()


def inin(x=1):
    if not x:
        return not x
    if x == 1:
        return int(inp())
    else:
        return map(int, inp().split())


def intarr():
    return list(inin(2))


def chrarr():
    return list(inp().split())


def lines(m):
    lines = [""] * m
    for i in range(m):
        lines[i] = list(inp().split())
    return lines


def intlines(m):
    lines = [0] * m
    for i in range(m):
        lines[i] = intarr()
    return lines


def pr(*a, sep=" ", end="\n", **d):
    h = list(d.keys())
    print(*a, *h, sep=sep, end=end)


def no(*a):
    pr("NO", *a)


def yes(*a):
    pr("YES", *a)


def No(*a):
    pr("No", *a)


def Yes(*a):
    pr("Yes", *a)


def ali(*a):
    pr("Alice", *a)


def bob(*a):
    pr("Bob", *a)


def pas(*a):
    pr("PASS", *a)


def pref(x):
    y = x.copy()
    for i in range(1, len(x)):
        y[i] += y[i - 1]
    return y


def suff(x):
    return pref(x[::-1])[::-1]


def binexp(a, b, res=1, mod=0):
    while b:
        if b & 1:
            res *= a
        a *= a
        b >>= 1
        if mod:
            a %= mod
            res %= mod
    return res


def test(t):
    while t:
        solve(t)
        t -= 1


def solve(t):
    n, q = inin(2)
    A = intarr()
    b = [inin() for _ in range(q)]
    res = [0] * 32
    corr = [0] * (32)
    z = 0
    for x in A:
        z |= x
    bit = 1 << 32
    cnt = 0
    for i in range(32, 0, -1):
        if z & bit:
            cnt += 1
        corr[i - 1] = cnt
        bit //= 2
    # Precompute the cost to set each bit level
    for i in range(32):
        a = A.copy()
        bit = 1 << i
        ans = 0
        for j in range(i, -1, -1):
            m = bit
            # Find minimum number of increments to get current power of two
            for x in a:
                m = max(min(bit - (x & (bit * 2 - 1)), m), 0)
            ans += m
            # Apply the increment to the first element matching required condition
            for k, x in enumerate(a):
                if bit - (x & (2 * bit - 1)) == m:
                    a[k] += m
                    break
            bit //= 2
        res[i] = ans
    # Binary search on prefix sums to find the best answer for each query
    for x in b:
        z = bisect_right(res, x)
        pr(z + corr[z])


test(inin(1))


# https://github.com/VaHiX/CodeForces/

File: CPP/ByTier/E/2147_E_Maximum_OR_Popcount.cpp
// Problem: CF 2147 E - Maximum OR Popcount
// https://codeforces.com/contest/2147/problem/E

// Problem: E. Maximum OR Popcount
// Task: Given an array of non-negative integers and multiple scenarios,
// each with a limit on operations (incrementing elements by 1), maximize
// the number of set bits in the bitwise OR of all numbers.
//
// Algorithm:
// - Precompute for all possible numbers of set bits (0 to 31) the minimum
//   number of operations needed to achieve that popcount.
// - Process each scenario by querying precomputed values to find max popcount
//   achievable within given operation limit.
//
// Time Complexity: O(N * 32 + Q * 32)
// Space Complexity: O(32) = O(1)

#include <algorithm>
#include <iostream>

using namespace std;
#define int long long
#define popc(x) __builtin_popcount(x)
const int N = 2e5 + 10, inf = 1e16 + 7;
int dp[60], a[N], n, q;

void work() {
  cin >> n >> q;
  for (int i = 1; i <= n; i++)
    cin >> a[i];
  for (int i = 0; i <= 50; i++)
    dp[i] = inf;
  int res = 0, val = 0;
  for (int i = 1; i <= n; i++)
    res |= a[i]; // Compute initial OR
  dp[popc(res)] = min(dp[popc(res)], val); // Store initial popcount cost
  for (int i = 0; i <= 31; i++) { // Iterate through all possible bit positions (0-31)
    for (int j = i; j >= 0; j--) {
      int f = 0;
      for (int k = 1; k <= n; k++)
        if (a[k] & (1ll << j))
          f = 1;
      if (f)
        continue; // Skip if bit already set in any element
      int pos = 0, ret = 0;
      for (int k = 1; k <= n; k++)
        if ((a[k] % (1ll << j)) >= ret) { // Find max value in current bit position
          pos = k;
          ret = (a[k] % (1ll << j));
        }
      val += (1ll << j) - ret; // Add cost to reach full bit set
      for (int k = 0; k < j; k++)
        if (a[pos] & (1ll << k))
          a[pos] ^= (1ll << k); // Clear lower bits
      a[pos] |= (1ll << j); // Set current bit
    }
    res = 0;
    for (int j = 1; j <= n; j++)
      res |= a[j]; // Recalculate OR after operations
    dp[popc(res)] = min(dp[popc(res)], val); // Store minimum cost to achieve popcount
  }
  while (q--) {
    int b;
    cin >> b;
    int ans = 0;
    for (int i = 0; i <= 31; i++)
      if (dp[i] <= b)
        ans = max(ans, i); // Find maximum popcount achievable with `b` operations
    cout << ans << "\n";
  }
  return;
}

signed main() {
  ios::sync_with_stdio(false);
  cin.tie(0);
  int T;
  cin >> T;
  while (T--)
    work();
  return 0;
}


// https://github.com/VaHiX/CodeForces/